src/shared/edit-util.c

```diff
@@ -221,67 +221,76 @@ static int run_editor(const EditFileContext *context) {
-                                _exit(EXIT_FAILURE);
-                        }
-                        n_editor_args = strv_length(editor_args);
-                        argc = n_editor_args;
-                }
-
-                argc += context->n_files * 2;
+                if (!isempty(editor)) {
+                        _cleanup_strv_free_ char **editor_args = NULL;
 
-                args = newa(char*, argc + 1);
+                        editor_args = strv_split(editor, WHITESPACE);
+                        if (!editor_args)
+                                goto oom;
 
-                if (n_editor_args > 0) {
-                        args[0] = editor_args[0];
-                        for (; i < n_editor_args; i++)
-                                args[i] = editor_args[i];
+                        r = strv_extend_strv(&args, editor_args, /* filter_duplicates = */ false);
+                        if (r < 0)
+                                goto oom;
                 }
 
                 if (context->n_files == 1 && context->files[0].line > 1) {
                         /* If editing a single file only, use the +LINE syntax to put cursor on the right line */
-                        if (asprintf(args + i, "+%u", context->files[0].line) < 0) {
-                                (void) log_oom();
-                                _exit(EXIT_FAILURE);
-                        }
+                        char l[3];
 
-                        i++;
-                        args[i++] = context->files[0].temp;
-                } else
-                        EDIT_FILES_FOREACH(f, *context)
-                                args[i++] = f->temp;
+                        xsprintf(l, "+%u", context->files[0].line);
 
-                args[i] = NULL;
+                        r = strv_extend(&args, l);
+                        if (r < 0)
+                                goto oom;
 
-                if (n_editor_args > 0)
+                        r = strv_extend(&args, context->files[0].temp);
+                        if (r < 0)
+                                goto oom;
+                } else
+                        EDIT_FILES_FOREACH(i, *context) {
+                                r = strv_extend(&args, i->temp);
+                                if (r < 0)
+                                        goto oom;
+                        }
+
+                if (!isempty(editor))
                         execvp(args[0], (char* const*) args);
 
+                bool prepended = false;
                 FOREACH_STRING(name, "editor", "nano", "vim", "vi") {
-                        args[0] = (char*) name;
-                        execvp(name, (char* const*) args);
+                        _cleanup_free_ char *e = NULL;
+
+                        e = strdup(name);
+                        if (!e)
+                                goto oom;
+
+                        if (!prepended) {
+                                r = strv_prepend(&args, e);
```

Comment:
`strv_prepend()` internally calls `strdup()`, hence the above `strdup()` is not necessary.